### PR TITLE
Refactor coagulation

### PR DESCRIPTION
**Pull request by huiwanpnnl**
The coagulation process in the aerosol microphysics package is refactored. Code changes are mainly in `modal_aero_coag.F90`:

- Subroutine `mam_coag_1subarea` is split into smaller subroutines; the contents not used by MAM4 are removed; many variables names are changed to (hopefully) more self-explanatory ones.
- Unused variables, data blocks, and calculations in subroutines `getcoags_wrapper_f` and `getcoags` are removed.
- The initialization of coagulation pairs in `modal_aero_amicphys:modal_aero_amicphys_init` is wrapped in a new subroutine `set_coagulation_pairs` in `modal_aero_coag.F90`; the contents not used by MAM4 are removed.

Please note that since the subroutine `mam_coag_1subarea` is significantly restructured, the comparison of `modal_aero_coag.F90` provided below by GitHub is not very helpful. To review the changes, it would be easier to simply open the new file in an editor window and read through it.

SMS test passed.

### CODE REVIEW

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 1]
```diff
@@ -1,2337 +1,1363 @@
...
+   integer :: ip
+
+   modefrm_coagpair(:) = big_neg_int
+   modetoo_coagpair(:) = big_neg_int
+   modeend_coagpair(:) = big_neg_int
+
+   n_coagpair = 3
+   ip=1; modefrm_coagpair(ip)=nait; modetoo_coagpair(ip)=nacc; modeend_coagpair(ip)=nacc
+   ip=2; modefrm_coagpair(ip)=npca; modetoo_coagpair(ip)=nacc; modeend_coagpair(ip)=nacc
+   ip=3; modefrm_coagpair(ip)=nait; modetoo_coagpair(ip)=npca; modeend_coagpair(ip)=nacc
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: I really like this! Thanks for removing that hard to follow do-loop. This is much easier to read!

File: components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90 [hunk 2]
```diff
@@ -4481,6 +4475,8 @@ subroutine modal_aero_amicphys_init( imozart, species_class,n_so4_monolayers_pca
     nspec_amode, &
     numptr_amode, numptrcw_amode, sigmag_amode
 
+use modal_aero_coag, only: set_coagulation_pairs
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90:4478]: (not related to this line) There is a use statement in this file at line 18:`use modal_aero_amicphys_control`. I think we should add `only` with that statement as it will make it easier to know which variable is coming from where. I know it will be a long list but I think there is value in adding that. We can add it in this PR or the next which will deal with the interfaces.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 3]
```diff
@@ -1,2337 +1,1363 @@
...
 !
-! !MODULE: modal_aero_coag --- modal aerosol coagulation
+! Modal aerosol coagulation parameterization.
 !
-! !INTERFACE:
-   module modal_aero_coag
+! Revision history:
+!   Richard C. Easter, 07.04.13:  Adapted from MIRAGE2 code. Some of the 
+!                                 code came from the CMAQ model v4.6.
+!   Hui Wan 2022: Refactored code: memoved non-4-mode codes; restructured subroutines.
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: memoved->removed

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 4]
```diff
@@ -1,2337 +1,1363 @@
...
+module modal_aero_coag
 
-! !USES:
-   use shr_kind_mod,    only:  r8 => shr_kind_r8
-   use cam_logfile,     only:  iulog
+  use shr_kind_mod,    only:  wp => shr_kind_r8
+  use cam_logfile,     only:  iulog
 
-   use modal_aero_amicphys_control
+  use modal_aero_amicphys_control
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Adding `only` for this use statement will make it easier to know which variables are coming form this module.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 5]
```diff
@@ -1,2337 +1,1363 @@
...
 
+contains
 
-!EOP
-!----------------------------------------------------------------------
-!BOC
-
-
-!EOC
+subroutine set_coagulation_pairs( big_neg_int )
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: In this case, I think we can get rid of `big_neg_int` variable and use something like the following within the subroutine:
```
integer, parameter:: huge_neg_int = -huge(huge_int)
```
and then use it like:
```
   modefrm_coagpair(:) = huge_neg_int
   modetoo_coagpair(:) = huge_neg_int
   modeend_coagpair(:) = huge_neg_int
```

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 6]
```diff
@@ -1,2337 +1,1363 @@
...
+      real(wp), intent(in) :: dgn_a(max_mode)
+      real(wp), intent(in) :: dgn_awet(max_mode) ! dry & wet geo. mean dia. (m) of number distrib.
+      real(wp), intent(in) :: wetdens(max_mode)  ! interstitial aerosol wet density (kg/m3)
+                                                 ! dry & wet geo. mean dia. (m) of number distrib.
 
-! *** modal geometric mean diameters: [ m ]
-      real(r8), intent(in) :: dgatk          ! aitken mode
-      real(r8), intent(in) :: dgacc          ! accumulation mode
+      real(wp), intent(inout), dimension(1:max_mode)           :: qnum_cur  ! current number mixing ratios (#/kmol-air)
+      real(wp), intent(inout), dimension(1:max_aer,1:max_mode) :: qaer_cur  ! current mass mixing ratios (kmol/kmol-air)
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: We have been putting units in square brackets like `[s]` for all the units. Would you please change the `()` to `[]` for these units?

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 7]
```diff
@@ -1,2337 +1,1363 @@
...
-! *** natural log of modal geometric standard deviation
-      real(r8), intent(in) :: xxlsgat         ! aitken mode
-      real(r8), intent(in) :: xxlsgac         ! accumulation mode
+      ! Local variables
 
-! *** coagulation coefficients
-      real(r8), intent(out) :: qs11, qn11, qs22, qn22,   &
-                               qs12, qs21, qn12, qv12
+      integer :: ip                 ! coagulation pair index
+      integer :: modefrm, modetoo   ! mode indices corresponding to the origin and desination of mass transfer
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: We have been using "src" and "dest" for origin/source and destination respectively. It would be nice to stay consistent.
```
integer :: src_mode, dest_mode
```

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 8]
```diff
@@ -1,2337 +1,1363 @@
...
+    real(wp), intent(in) :: xxlsgac         ! accumulation mode
+
+    ! Coagulation coefficients
+    real(wp), intent(out) :: qn11, qn22, qn12, qv12 
+
+    integer ibeta, n1, n2a, n2n ! indices for correction factors
+
+    real(wp) :: kngat, kngac
+    real(wp),parameter :: one = 1.0_wp, two = 2.0_wp, half = 0.5_wp
+    real(wp),parameter :: a = 1.246_wp
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: We are trying to get rid of single letter variables. If you can change `a` to something meaningful or just `aa` or if it is a coefficient, `a_coef` or `coef_a`, that will be great.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 9]
```diff
@@ -1,2337 +1,1363 @@
...
+    real(wp) esac01         ! accumulation mode exp( log^2( sigmag )/8 )
+
+    real(wp) esat04,esat05,esat08,esat09,esat16,esat20,esat24,esat25,esat36,esat49,esat64,esat100
+    real(wp) esac04,esac05,esac08,esac09,esac16,esac20,esac24,esac25,esac36,       esac64
+
+    real(wp) dgat2, dgac2, dgat3, dgac3
+    real(wp) sqdgat, sqdgac
+    real(wp) sqdgat5, sqdgac5
+    real(wp) sqdgat7
+    real(wp) r, r2, r3, rx4
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Same is the case here with variable `r`.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 10]
```diff
@@ -1,2337 +1,1363 @@
...
+    rat = dgacc / dgatk
+
+    ! Trap subscripts for bm0 and bm0i, between 1 and 10.
+    ! See page h.5 of whitby et al. (1991)
+
+    n2n = max( 1, min( 10,   nint( 4.0_wp * ( sgatk - 0.75_wp ) ) ) )
+    n2a = max( 1, min( 10,   nint( 4.0_wp * ( sgacc - 0.75_wp ) ) ) )
+    n1  = max( 1, min( 10,   1 + nint( dlgsqt2 * log( rat ) ) ) )
+
+    !----------------------------------------------------
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Can we create a new subroutine for the following calculations in "Intermodal coagulation, set up for zeroeth moment"?

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 11]
```diff
@@ -1,2337 +1,1363 @@
...
-               qaer_tmpb(iaer,nait ) = qaer_tmpb(iaer,nait ) + tmp_dq*tmp3
-               qaer_tmpb(iaer,nmacc) = qaer_tmpb(iaer,nmacc) + tmp_dq*tmp4
-               qaer_del_coag_in(iaer,i_agepair_pca ) &
-                                    = qaer_del_coag_in(iaer,i_agepair_pca ) + tmp_dq*tmp2
-               qaer_del_coag_in(iaer,i_agepair_macc) &
-                                    = qaer_del_coag_in(iaer,i_agepair_macc) + tmp_dq*tmp4
-            end do
-         end if
-      end if
+    !-----------------------------------------------------------------
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Same here, a new subroutine for third moment.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 12]
```diff
@@ -1,2337 +1,1363 @@
...
-      end do
-      end if
-#endif
+    qv12 = coagnc3 * coagfm3 / ( coagnc3 + coagfm3 )
 
-      qnum_cur = qnum_tmpb
-      qaer_cur = qaer_tmpb
+    !-----------------------------------------------------
+    ! Intramodal coagulation
+    !-----------------------------------------------------
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: The following calculations for aitken and acumulation modes are very similar. Can we create a subroutine and call it twice (for aitken and accumulation)with different inputs ?

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 13]
```diff
@@ -1,2337 +1,1363 @@
...
 
-      qn12 = coagatac0
+      qnum_tavg(npca) = (qnum_bgn(npca) + qnum_end(npca))*0.5_wp
 
+      !-----------------------------------------------------------------------------------------
+      ! aitken mode number loss - approximate analytical solution
+      ! using average number conc. for accum, pcarbon, and marine-org accum modes
+      !-----------------------------------------------------------------------------------------
+      tmpa = ybetaij0(1)*qnum_tavg(nacc)
+      tmpa = tmpa + ybetaij0(3)*qnum_tavg(npca)
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Can we rename `tmpa` to something more meaningful? It is okay to use `tmp` but can add a string like `tmp_avg` or something similar?

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 14]
```diff
@@ -1,2337 +1,1363 @@
...
+! Purpose: update the number mixing ratio of a single mode considering self-coagulation 
+!          and inter-modal coagulation
+!-----------------------------------------------------------------------------------------
 
+    real(wp),intent(in)  :: rateij, rateii  ! inter-modal and self-coagulation rates
+    real(wp),intent(in)  :: qnum_bgn        ! qnum (number mixing ratio) start value
+    real(wp),intent(out) :: qnum_end        ! qnum (number mixing ratio) end value
 
-! *** loss to accumulation mode
+    real(wp) :: tmpc
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Can we rename `tmpc` to `exp_rate` or `tmp_exp` or something similar?

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 [hunk 15]
```diff
@@ -1,2337 +1,1363 @@
...
-              + two * r2 * esat25  * esac04   &
-              + rx4 * esat09  * esac16   &
-              + ri3 * esat100 * esac09   &
-              + two * ri1 * esat64  * esac01 )
+      rate1 = max( 0.0_wp, ybetaij3(1)*qnum_tavg(nacc) )
+      rate2 = max( 0.0_wp, ybetaij3(3)*qnum_tavg(npca) )
+      ratesum = rate1 + rate2
 
-! *** gain by accumulation mode = loss from aitken mode
+      tmpc = deltat*ratesum ! calculate coag-induced changes only when this number is not ~= zero
```
Comment (singhbalwinder) [inline components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Can we rename `tmpc` to `tmp_coag` or something similar?

**Review by singhbalwinder — CHANGES_REQUESTED** [review]
Thanks, Hui! This is very neatly written code. I really liked that way you factored out codes. It makes it a lot easier to read. I have suggested few changes. Let me know if you have any comments or questions on these changes.

File: components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90 (hunk 2 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90:4478]: There is another comment below about `use modal_aero_amicphys_control` in `modal_aero_coag.F90`. I've addressed that one by changing to `use ... only: ...`.

This comments, in contrast, is about `modal_aero_amicphys.F90`. Since `modal_aero_amicphys_control.F90` contains pretty much all the module variables originally declared in `modal_aero_amicphys.F90`, I think in a `use ... only: ...` statement we will end up having to list everything (i.e., 100+ variables :-) ). I would suggest we take the second option mentioned above by @singhbalwinder, i.e., defer this to a future PR. Also note that in the box model, we have split those 100+ module variables into groups related to the aerosol modes, species, individual processes, and the mapping of tracer indices between `amicphys` and the host model. It might be useful to discuss (for a future PR) whether we want to port those changes back to the global model.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 3 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Corrected.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 4 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Done. I also moved some additional coagulation-specific contents from `modal_aero_amicphys.F90` and `modal_aero_amicphys_control.F90` to this module.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 5 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: I've changed `big_neg_int` from an intent(in) argument to a local parameter:

```
integer, parameter :: big_neg_int = -huge(selected_int_kind(6))
```

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 7 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: I can make these changes. 

Do you want the following arrays to be renamed, too? 

```
  integer, public :: modefrm_coagpair(max_coagpair)
  integer, public :: modetoo_coagpair(max_coagpair)
  integer, public :: modeend_coagpair(max_coagpair)
```

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 6 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Yes, I'll go through the module and make the changes.

File: components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90 (hunk 2 above)
Comment (singhbalwinder) [reply components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90:4478]: Let's defer that to the interface PR. For the rename process, I also saw that some variables (e.g. log of `sigmag`) are redefined in amicphys for all the modes. There may be other similar calculations that we can directly use from modal_aero_data.F90 or other places in the model. This will help us bring down the number of variables in amicphys_control

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 5 above)
Comment (singhbalwinder) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: There is a typo in my suggestion above. We can do something like this:
`integer, parameter:: huge_neg_int = -huge(huge_neg_int)`

What you used above is also fine but the above is a more typical use of `huge`.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 7 above)
Comment (singhbalwinder) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Yes, I think we can rename them as `src_mode_coagpair`, `dest_mode_coagpair`  or something similar

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 5 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Out of curiosity: in your suggested line of code, we have `huge_neg_int` on both the LHS and the RHS of the assignment statement. Isn't this recursive, or did I miss some language feature in Fortran?

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 7 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: OK. Will do!

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 5 above)
Comment (singhbalwinder) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: It looks a bit odd but it is not recursive. Fortran first declares the variable of a particular type (e.g. integer, real, etc.) and then uses that type to assign a value to `huge`. It is better this way as then we don't need to go look what `select_int_kind(6)` is. We just know that the variable is assigned a `huge` value of its own type.

Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Ah, good to know. I'll update the code following your suggestion.

**Comment by singhbalwinder** [comment]
BTW, the `fast_compile_run.sh` script now works. It will make it faster to build the model. Since you are using a version without this fix, you can just remove space before `  /share/...` in the LD_LIBRARY_PATH variable in the `.env_mach_specific.sh` file (it is a hidden file) in the case directory. After that execute` bash fast_compile_run.sh` to build and submit a run.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 8 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Since `a` is a constant, I renamed it to `a_const`.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 9 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Changed to `r1` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/b13d53649eec63411da011f54c7f43184a0b907e).

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 11 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Done in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/15badfdd0446e21d10323d34d36f00f93dc2858e).

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 10 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Done in commit [13eb6a4](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/13eb6a4ea737604b67dddeb350f6b1facb7b306c)

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 12 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Good point :-)  - and done in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/060e49d6375fc68873b3ae45cf99bbb0f4199d6c)

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 13 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: I changed `tmpa` to `bijqnumj` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/0aacb82f99405c7f4207680de523988fb31ce395) and explained the meaning in the comments: it's the intermodal coagulation coefficient bij multiplied by the number mixing ratio of mode j. 

Some of the variables in these low-level subroutines are hard to give good names to, because they correspond to mathematical expressions that are specific to the numerical algorithms in use and they don't have a clear physical meaning. By using the variable name `bijqnumj` for what used to be `tmpa`, I'm trying to spell out what the quantity being calculated is, which can be useful for readers who want to check or understand the numerical algorithm. Hope this is acceptable.

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 14 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: Changed to `tmp_exp` in [this commit](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/0aacb82f99405c7f4207680de523988fb31ce395).

File: components/eam/src/chemistry/modal_aero/modal_aero_coag.F90 (hunk 15 above)
Comment (huiwanpnnl) [reply components/eam/src/chemistry/modal_aero/modal_aero_coag.F90]: In [this commit]([607f597](https://github.com/eagles-project/e3sm_mam4_refactor/pull/32/commits/607f597a3134a0a3aafbfc3b11ed655562404c71)), I revised the comments in the subroutine to mention that the approximated mass mixing ratio equations take the form of exponential decay equations. In that context, I changed `ratesum` to `decay_const` and `tmpc` to `decay_factor`.

**Comment by huiwanpnnl** [comment]
Hi @singhbalwinder, I have responded or addressed all your comments above. Please take a look and let me know if additional changes are needed.

This branch now has a minor conflict with a recent update on `refactor-maint-2.0`, which resulted in the `mam4_refactor_scripts/compare_baselines_all_tests.sh` script to give an "all-fail" result. This is because the `git merge` command in the script inserted markings of merge conflicts in `modal_aero_amicphys.F90`, which, as expected, caused `case.build` to fail. I have resolved the conflicts, pushed the updated code to a new branch, and issued a new PR (#42) to replace this one.

**Comment by singhbalwinder** [comment]
Replaced by #42
